app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java

```diff
@@ -0,0 +1,183 @@
+            throw new IllegalArgumentException("location provider cannot be null or empty");
+        }
+
+        mProviders.add(locationProvider);
+    }
+
+
+    /**
+     * Registers fetcher for location updates
+     *
+     * @param minSamplingFrequencyMs         - minimum time between location samples,  in milliseconds
+     * @param minDistanceDeltaSamplingMeters - minimum distance between location samples, in meters
+     */
+    public void registerForUpdates(long minSamplingFrequencyMs,
+                                   float minDistanceDeltaSamplingMeters) {
+        if (minSamplingFrequencyMs < 0) {
+            throw new IllegalArgumentException("minSamplingFrequencyMs cannot be negative");
+        }
+
+        if (minDistanceDeltaSamplingMeters < 0) {
+            throw new IllegalArgumentException(
+                    "minDistanceDeltaSamplingMeters cannot be a negative number");
+        }
+
+        if (mIsRegistered) {
+            throw new RuntimeException("Fetcher already registered!");
+        }
+
+        checkForLocationPermission(mContext);
+
+        if (mProviders.isEmpty()) {
+            throw new RuntimeException("No providers added to fetcher to register with");
+        }
+
+        for (String provider : mProviders) {
+            mLocationManager.requestLocationUpdates(provider, minSamplingFrequencyMs,
+                    minDistanceDeltaSamplingMeters, mLocationListener);
+        }
+
+        mIsRegistered = true;
+    }
+
+    /**
+     * Stops fetcher from receiving location updates
+     */
+    public void unregisterFromUpdates() {
+        if (!mIsRegistered) {
+            throw new RuntimeException("Fetcher is not registered");
+        }
+
+        mLocationManager.removeUpdates(mLocationListener);
+        mIsRegistered = false;
+    }
+
+
+    /**
+     * Used for receiving notifications from the {@link LocationFetcher}
+     * when location events occurs.
+     * New location notifications will be raised only after the
+     * {@link LocationFetcher} registered for location changes
+     */
+    public interface LocationFetcherListener {
+        void onProviderDisabled(@ProviderType String locationProvider);
+
+        void onProviderEnabled(@ProviderType String locationProvider);
+
+        void onNewLocationSample(LocationSample locationSample);
+    }
+
+    /**
+     * Encapsulates incoming updates handling from {@link LocationFetcher}'s underlying {@link LocationManager}
+     */
+    private class MyLocationListener implements LocationListener {
+
+        @Override
+        public void onLocationChanged(Location location) {
+            LocationSample locationSample = new LocationSample(location);
+            mLocationFetcherListener.onNewLocationSample(locationSample);
+        }
+
+
+        @Override
+        public void onStatusChanged(String provider, int status, Bundle extras) {
+            if (status == android.location.LocationProvider.AVAILABLE) {
```

Comment:
android.location.LocationProvider hold a few static ints that represent the status result domain.
Available is the only one that means the provider is enabled.
Just to make it clear, it's the Android LocationListener that im wrapping. A user of LocationFetcher isn't aware of this status.